src/main/java/redis/clients/jedis/csc/CacheConnection.java

```diff
@@ -62,30 +63,39 @@ public void disconnect() {
 
   @Override
   public <T> T executeCommand(final CommandObject<T> commandObject) {
-    CacheKey key = new CacheKey<>(commandObject);
-    if (!clientSideCache.isCacheable(key)) {
+    final CacheKey<T> cacheKey = new CacheKey<>(commandObject);
+
+    // Check if key is cachable at all, if not just execute the command as if there is no cache enabled
+    if (!clientSideCache.isCacheable(cacheKey)) {
       clientSideCache.getStats().nonCacheable();
       return super.executeCommand(commandObject);
     }
 
-    final CacheKey cacheKey = new CacheKey(commandObject);
+    // Check if the cache contains an entry for the provided key
     CacheEntry<T> cacheEntry = clientSideCache.get(cacheKey);
-
-    // CACHE HIT !!
+    cacheEntry = validateEntry(cacheEntry);
     if (cacheEntry != null) {
-      cacheEntry = validateEntry(cacheEntry);
-      if (cacheEntry != null) {
-        clientSideCache.getStats().hit();
-        return (T) cacheEntry.getValue();
-      }
+      // We only return the key if it is valid, otherwise we follow the flow if there was no cache hit
+      clientSideCache.getStats().hit();
+      return (T) cacheEntry.getValue();
     }
 
-    // CACHE MISS!!
+    // ---
+    // At this point we know there is no valid cache entry to return,
+    // so we attempt to pull one from the server and cache it
+    // ---
+
     clientSideCache.getStats().miss();
+    cacheEntry = CacheEntry.inProgress(cacheKey, this);
+    clientSideCache.set(cacheKey, cacheEntry);
+
     T value = super.executeCommand(commandObject);
-    if (value != null) {
-      cacheEntry = new CacheEntry<>(cacheKey, value, this);
+    cacheEntry = clientSideCache.get(cacheKey);
+    if (shouldCacheEntry(value, cacheEntry)) {
+      cacheEntry = CacheEntry.newCacheEntry(cacheKey, this, value);
```

Comment:
I don't this is properly doing:

> If the correct dummy entry is still there, replace it with the actual value.

The correct dummy entry is the one we used in previous `clientSideCache.set(cacheKey, cacheEntry);` but current `cacheEntry` is not guaranteed to be that one.
Furthermore, it has to be ensured to replace THAT one.